agent/credentialssource/credentials_source.go

```diff
@@ -0,0 +1,196 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package credentialssource provides managing of defaults file.
+package credentialssource
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/ini.v1"
+
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/percona/pmm/api/inventorypb"
+)
+
+// Parser is a struct which is responsible for parsing credentialsJSON source/defaults file.
+type Parser struct{}
+
+// New creates new Parser.
+func New() *Parser {
+	return &Parser{}
+}
+
+type credentialsSource struct {
+	username      string
+	password      string
+	host          string
+	agentPassword string
+	port          uint32
+	socket        string
+}
+
+// credentialsJSON provides access to an external provider so that
+// the username, password, or agent password can be managed
+// externally, e.g. HashiCorp Vault, Ansible Vault, etc.
+type credentialsJSON struct {
+	AgentPassword string `json:"agentpassword"`
+	Password      string `json:"password"`
+	Username      string `json:"username"`
+}
+
+// ParseCredentialsSource parses given file in request. It returns the database specs.
+func (d *Parser) ParseCredentialsSource(req *agentpb.ParseCredentialsSourceRequest) *agentpb.ParseCredentialsSourceResponse {
+	var res agentpb.ParseCredentialsSourceResponse
+	parsedData, err := parseCredentialsSourceFile(req.FilePath, req.ServiceType)
+	if err != nil {
+		res.Error = err.Error()
+		return &res
+	}
+
+	res.Username = parsedData.username
+	res.Password = parsedData.password
+	res.AgentPassword = parsedData.agentPassword
+	res.Host = parsedData.host
+	res.Port = parsedData.port
+	res.Socket = parsedData.socket
+
+	return &res
+}
+
+func parseCredentialsSourceFile(filePath string, serviceType inventorypb.ServiceType) (*credentialsSource, error) {
+	if len(filePath) == 0 {
+		return nil, errors.New("configPath for parseCredentialsSourceFile is empty")
+	}
+
+	filePath, err := expandPath(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("fail to normalize path: %w", err)
+	}
+
+	// check if file exist
+	if _, err := os.Stat(filePath); errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Errorf("file doesn't exist: %s", filePath)
+	}
+
+	credentialsJSONFile, err := parseJsonFile(filePath)
+	if err == nil {
+		return credentialsJSONFile, nil
+	}
+
+	if serviceType == inventorypb.ServiceType_MYSQL_SERVICE {
+		return parseMySQLDefaultsFile(filePath)
+	}
+
+	return nil, errors.Errorf("unrecognized file type %s", filePath)
+}
+
+func parseJsonFile(filePath string) (*credentialsSource, error) {
+	creds := credentialsJSON{"", "", ""}
+
+	f, err := os.Lstat(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
```

Comment:
let's use `errors.Wrapf` in all cases below with `%w`

Reply:
Changed.